docs: add CIP for block limits on pfbs and tx size (#220)
celestiaorg/CIPs: cips/cip-tx-limits.md

| cip | TBD(27) |
| - | - |

> | title | Block limits for number of PFBs/non-PFBs and transaction size |

Reviewer: We discussed this in all core devs but IMO there wasn't a clear decision made in the call. I think we could split this into two CIPs:

1. Block limits for PFBs / non-PFBs 
2. Limit transaction size

Author: I think @cmwaters explanation is that these both impact how a block is constructed, so was okay to leave it as 1 CIP.

Author: I'm also happy to scrap this and have it split to 2 CIPs by authors of the features.

> | title | Block limits for number of PFBs/non-PFBs and transaction sizes |

Author: ```suggestion
| title | Block limits for number of PFBs/non-PFBs and transaction sizes |
```

| description | Set limits for number of PFBs and non-PFBs per block and transaction size |
| author | Josh Stein ([@jcstein](https://github.com/jcstein)), Nina Barbakadze ([@ninabarbakadze](https://github.com/ninabarbakadze)) |
| discussions-to |  <https://forum.celestia.org/t/cip-limit-number-of-pfbs-and-non-pfbs-per-block-increase-transaction-size-limit/1843> |
| status | Draft |
| type | Standards Track |
| category | Core |
| created | 2024-10-16 |

## Abstract

> This CIP proposes to set limits for the number of PFBs and non-PFBs per block and set the limit for transaction size. Concretely, the proposal is to set the limits to 600 PFBs and 200 non-PFB transactions per block and set the transaction size limit to 2MiB.

Author: https://github.com/celestiaorg/CIPs/pull/220/commits/164e63c336e22012c7fe8ad29c6ef202613fc83b

## Specification

> 1. The number of PFBs per block is limited to 600 by setting `MaxPFBMessages` to 600. `MaxPFBMessages` is the maximum number of PFB messages a block can contain.

Author: ```suggestion
1. The number of PFBs per block is limited to 600 by setting `MaxPFBMessages` to 600.
```

> 1. The number of non-PFB transactions per block is limited to 200 by setting `NonPFBTransactionCap` to 200. `NonPFBTransactionCap` is the maximum number of SDK messages, aside from PFBs, that a block can contain.

Reviewer: [optional] I think it's worth calling out that these limits aren't actually enforced. They're limits used by the celestia-app implementation but a validator could hypothetically modify prepare proposal, run that binary and create blocks with more PFBs or non-PFBs.

[nit] the constants are poorly named b/c the thing being limited is the number of messages in a block not the number of transactions. Should we change the implementation and the specs here? cc: @rach-id

Author: https://github.com/celestiaorg/CIPs/pull/220/commits/2d60926e4786ebed5c909c8176332bef340c4609

Reviewer 2: @rootulp yes sure. want me to open a PR or you do it?

Author: https://github.com/celestiaorg/CIPs/pull/220/commits/8fbee0ed2f37bc2c8c97be877dc007c92b85bf55 updated to new variables, if these don't get included with v3.0.0-rc, I will revert to names `NonPFBTransactionCap` and `PFBTransactionCap`

> 

Reviewer 3: ```suggestion

It's important to note that these limits are not strictly enforced. While they are defined by the celestia-app implementation, a validator could potentially modify the prepareProposal logic, run a custom binary, and produce blocks that exceed the specified limits for PFB or non-PFB transactions.

```

> 1. The size of a transaction is limited to 2MB by setting `MaxTxBytes` to 2091752, which is 2MiB in bytes.

Reviewer: typo in the number of bytes

the constant was renamed to MaxTxSize

```suggestion
1. The size of a transaction is limited to 2MiB by setting `MaxTxSize` to 2097152, which is 2MiB in bytes.
```

Reviewer: Optional I think this CIP could describe the new constant `MaxTxSize` and how it is enforced (via an antehandler). cc: @ninabarbakadze

Author: https://github.com/celestiaorg/CIPs/pull/220#discussion_r1803346789

> 1. The size of a transaction is limited to 2MB by setting `MaxTxBytes` to 2091752, which is 2MiB in bytes.

Reviewer 3: ```suggestion
1. The size of a transaction is limited to 2MB by setting `MaxTxBytes` to 2091752, which is 2MiB in bytes. From version v3 and above, in `CheckTx`, `PrepareProposal`, and `ProcessProposal`, each transaction's size is checked against the `appconsts.MaxTxSize` threshold. This ensures that transactions over the limit are rejected or excluded at all stages, from initial submission to execution.
```

Author: https://github.com/celestiaorg/CIPs/pull/220/commits/f401653d08accfadc659e4d3b7ebe7bdc76f0123 added here because suggestion wasn't mergable, as it was on outdated version of PR

> 1. The number of non-PFBs messages per block is limited to 200 by setting `MaxNonPFBMessages` to 200. `MaxNonPFBMessages` is the maximum number of SDK messages, aside from PFBs, that a block can contain.

Reviewer 4: This feels a bit verbose

Author: reducing, thank you

> 1. The number of non-PFBs messages per block is limited to 200 by setting `MaxNonPFBMessages` to 200. `MaxNonPFBMessages` is the maximum number of SDK messages, aside from PFBs, that a block can contain.

Author: ```suggestion
1. The number of non-PFBs messages per block is limited to 200 by setting `MaxNonPFBMessages` to 200.
```

## Rationale

The rationale for this proposal is twofold:

> 1. To prevent congestion on the network by limiting the number of PFBs and non-PFB transactions per block. This was initially not considered consensus-breaking, but it has a meaningful effect on users and should be formalized in a CIP.

Reviewer: [optional] I think specifically to prevent block times from spiking b/c if a block includes > 600 PFB messages and 200 non-PFB messages, the time to process the block expands past the 6 second time window. cc: @rach-id can clarify

```suggestion
1. To prevent long block times on the network by limiting the number of PFBs and non-PFB transactions per block. This was initially not considered consensus-breaking, but it has a meaningful effect on users and should be formalized in a CIP.
```

> 1. Transaction size is limited to 2MiB by setting `MaxTxSize` to 2097152, which is 2MiB in bytes. From version v3 and above, in `CheckTx`, `PrepareProposal`, and `ProcessProposal`, each transaction's size is checked against the `appconsts.MaxTxSize` threshold. This ensures that transactions over the limit are rejected or excluded at all stages, from initial submission to execution.

Author: ```suggestion
1. Transaction size is limited to 2MiB by setting the versioned parameter`MaxTxSize` to 2097152, which is 2MiB in bytes. From version v3 and above, in `CheckTx`, `PrepareProposal`, and `ProcessProposal`, each transaction's size is checked against the `appconsts.MaxTxSize` threshold. This ensures that transactions over the limit are rejected or excluded at all stages, from initial submission to execution.
```

Author: cc @cmwaters

> 1. The size of a transaction is limited to 2MiB by setting `MaxTxSize` to 2097152, which is 2MiB in bytes. From version v3 and above, in `CheckTx`, `PrepareProposal`, and `ProcessProposal`, each transaction's size is checked against the `appconsts.MaxTxSize` threshold. This ensures that transactions over the limit are rejected or excluded at all stages, from initial submission to execution.

Author: ```suggestion
1. Transaction size is limited to 2MiB by setting `MaxTxSize` to 2097152, which is 2MiB in bytes. From version v3 and above, in `CheckTx`, `PrepareProposal`, and `ProcessProposal`, each transaction's size is checked against the `appconsts.MaxTxSize` threshold. This ensures that transactions over the limit are rejected or excluded at all stages, from initial submission to execution.
```

1. To set the transaction size limit to 2MiB, even with 8MB blocks, to prevent issues with gossiping large transactions. Gossiping an 8MiB transaction without chunking could be detrimental to the network.

> These changes affect users significantly and are consensus-breaking, necessitating their inclusion in a CIP.

Reviewer: [note] only the `MaxTxBytes` change is consensus breaking. The other one isn't.

Author: https://github.com/celestiaorg/CIPs/pull/220/commits/c6d9a7c71a16044bb92d11caf569039603d14d8b

> 1. The size of a transaction is limited to 2MiB by setting `MaxTxSize` to 2097152, which is 2MiB in bytes. From version v3 and above, in `CheckTx`, `PrepareProposal`, and `ProcessProposal`, each transaction's size is checked against the `appconsts.MaxTxSize` threshold. This ensures that transactions over the limit are rejected or excluded at all stages, from initial submission to execution.

Reviewer 4: This first line also feels verbose. We should also point out that this is a versioned parameter

Author: https://github.com/celestiaorg/CIPs/pull/220#discussion_r1805085934

> 1. To prevent long block times on the network by limiting the number of PFBs and non-PFBs messages per block. This was initially not considered consensus-breaking, but it has a meaningful effect on users and should be formalized in a CIP.

Author: ```suggestion
1. To prevent long block times on the network by limiting the number of PFBs and non-PFBs messages per block. This was initially not considered consensus-breaking, but it has a meaningful effect on users and should be formalized in a CIP. 
    1. The limits for PFBs (Pay for Blob transactions) and non-PFBs per block were established using the following process:
        1. Benchmarks were conducted in [PR 3904 on celestia-app](https://github.com/celestiaorg/celestia-app/pull/3904) to measure ABCI method processing times for different transaction types.
        1. A target processing time of ~0.25 seconds was set to prevent long block times.
        1. Based on these benchmarks, run on the recommended validator configuration (4 CPU, 16GB RAM), a soft limiter was implemented in the prepare proposal stage.
        1. This limiter sets specific caps on the number of PFB and non-PFB messages allowed in a default block to meet the processing time target.
        1. While default blocks adhere to these limits, blocks exceeding them can still be included if they reach consensus, ensuring flexibility.
    1. This approach balances network efficiency with block processing speed, directly informing the PFB and non-PFB limits now in place.
```

cc: @rach-id

> 1. To prevent long block times on the network by limiting the number of PFBs and non-PFBs messages per block. This was initially not considered consensus-breaking, but it has a meaningful effect on users and should be formalized in a CIP.

Reviewer 4: We can do this at a later point but it might be helpful to remember later on how these numbers were chosen

Reviewer 2: we could point to this PR https://github.com/celestiaorg/celestia-app/pull/390 as I am adding the results to it and will open it for review afterwards

Author: The PR 309 referenced is from deciding on the prefix for addresses and unrelated to this? https://github.com/celestiaorg/celestia-app/issues/390
<img width="349" alt="Screenshot 2024-10-17 at 12 26 35" src="https://github.com/user-attachments/assets/f1fe924e-f910-4c2b-b487-c54a55111b15">

@rach-id Can you please provide a condensed explanation of what you learned in the PR that was intended to be linked?

Reviewer 2: my bad, this PR: https://github.com/celestiaorg/celestia-app/pull/3904

That PR contains benchmarks that we run to measure the processing time different ABCI methods take with different transaction types.

Then, to limit that processing time to ~0.25sec, we added a soft limiter in prepare proposal so that the default block created by the block construction mechanism respects that processing time limit. 

This soft limiter is a simple cap on the number of messages a block can contain, based on running the PRs benchmarks on 4 CPU 16GB RAM machines, which is the suggested validator configuration.

This way, the default blocks don't take much time to process. However, if a block not respecting those limits reaches consensus, it still can be included and won't be wasted.

## Backwards Compatibility

> This proposal is meant to be included with v3 and the [Ginger Network Upgrade](./cip-25.md). It is not backwards compatible with v2.

Reviewer 4: The feature itself is actually backwards compatible because it's not a consensus breaking change

> This proposal is meant to be included with v3 and the [Ginger Network Upgrade](./cip-25.md). It is not backwards compatible with v2.

Author: ```suggestion
This proposal is meant to be included with v3 and the [Ginger Network Upgrade](./cip-25.md). It is backwards compatible with v2.
```
